### ISSUE TITLE
[Feature Request] Enable cli to use allready open instance of fork (single instance)

### ISSUE THREAD

**Issue by felixhirt** [ForkIssues/TrackerWin#266]
It would be awesome if i could tell fork from the cli to open a given repo in the allready open instance (or start up, if fork.exe is not allready running), something like fork.exe --singleInst "path/to/my/repo"

**Comment by felixhirt** [comment]
This could also be used as ground work for #89

**Comment by felixhirt** [comment]
Just a quick follow up: Was this implemented wIth the release of 1.37? And if so, is there a doc for the cli arguments?

**Comment by DanPristupov** [comment]
No, sorry. I'll do my best to introduce it in 1.38.

The task is 50% done, but not everything works properly yet.

This feature seems to be trivial, but it requires IPC communication which is always difficult and error prone.

**Comment by felixhirt** [comment]
Thanks for the quick response and all the work you put into fork. I only asked because it is still in the 1.37 milestone, no problem if it takes longer.
PS: I agree, IPC is hard, especially if you dont control the rest of the environment

**Comment by DanPristupov** [comment]
I'm uploading 1.38 update right now. It will handle CLI requests in the active instance (if it exists).

**Comment by felixhirt** [comment]
Thanks a lot. Is there a documentation somewhere for the cli arguments possible? i tried fork -help and fork /?. I can open a second instance with "fork" or just focus the current instance with "fork --singleinst", but i cannot open a repo in the current instance

**Comment by DanPristupov** [comment]
Currently the only option is the repository path.

examples: 
`fork .`
`fork c:\path\to\repo`

It should find an active instance (or create a new one if none were found) and handle the request in it.

I think second instance should never be opened (if you are using Fork higher than 1.38 version, of course). Does this really happen?

**Comment by lnist** [comment]
From PowerShell `C:\Users\<my.user>\AppData\Local\Fork\Fork.exe $pwd` works and opens in already open instance, but `C:\Users\<my.user>\AppData\Local\Fork\Fork.exe .` only focuses the open instance.

**Comment by DanPristupov** [comment]
It looks like the working directory path is incorrect when command is called from Powershell. So, relative paths don't work, however absolute paths must be handled properly.

At the time I'm trying to figure out why Fork gets a wrong working directory.

**Comment by DanPristupov** [comment]
Bad news. This is a bug in Squirrel (https://github.com/Squirrel/Squirrel.Windows) which doesn't pass working directory into the main process and makes the relative path handling impossible.

I don't see any possible fix at the time.

:(

**Comment by felixhirt** [comment]
I see. i guess it tried to open the repo "--singleinst" in my previous attempts. If i call it with no arguments at all, it opens a new instance, if i call it with an absolute path, it opens the repo in the running process. super awesome!
The problem with relative paths is unfortunate, especially because it is difficult to document. But i would assume it did not work before either? if so, i would suggest opening a new issue

**Comment by DanPristupov** [comment]
Yes, it looks like relative paths never really worked.

There's a 1 year old pull request with a workaround in the Squirrel repo, but it's still not approved. So, I guess we won't see a fix soon.

**Comment by felixhirt** [comment]
A few additions:
- There is a small problem: if a repo which is not allready open in fork is opened in the same instance, the tab is not selected (the repo is added and the tab opened, but not selected)
- For some reason it does not work anymore for me with the main "Fork.exe" (Squirrel executable). If i start the one in app-138.0 directly, it works... Any idea what i did wrong? I could have sworn it worked when i tested it yesterday (???)
I can reopen the issue if you want

**Comment by felixhirt** [comment]
in regards to the first point above: maybe make it so that if the app does not have focus, the tab is switched, but not otherwise (to prevent a script or something running in the background from interfering with the user)

**Comment by DanPristupov** [comment]
> There is a small problem: if a repo which is not allready open in fork is opened in the same instance, the tab is not selected (the repo is added and the tab opened, but not selected)

Hi. I will investigate that and answer a bit later.

To understand the problem better, could you describe how you use this feature?

**Comment by quotidian-ennui** [comment]
Ok, since I spend my life in git-bash I have a function that starts fork from the bash prompt

```
fork() {
  "/c/Users/<me>/AppData/Local/Fork/Fork.exe" $(cygpath -m $(pwd)) &
}
```

If I run that repeatedly from the same directory

1. No Fork Open -> creates a new Window
1. I switch to the fork window; switch back to the bash prompt run `fork` again
1. fork on the taskbar flashes; and there's only 1 window
1. Stay in the bash prompt and run fork again.
1. a new fork window is opened; so now I have two fork windows both with the same repo open.

The "new fork window" doesn't always happen, but if I execute fork enough times without switching away from git-bash, then multiple windows will appear.

I can live with that behaviour, since it fulfils my primary desire is to try to re-use the existing instance. but I thought I'd mention it, because it seems it might be related to the last issue around this feature

**Comment by felixhirt** [comment]
@DanPristupov I use it (or plan to use it at least) from my little helper tool which lists all my repos on a remote and allows me to quickly clone a selected one into the folder where it would belong on my local machine (We have lots of repos). after this is done, i would like to add it to fork for convenience. For that it would be helpfull if the tab would be selected in fork.
Right now, i use it by opening a cmd console in the install directory of fork and run something like "fork path/to/my/repo" (for testing)

**Comment by felixhirt** [comment]
@quotidian-ennui does it work allways if you use "C:\Users\<me>\AppData\Local\Fork\app-1.38.0\Fork.exe"? because like that, it works for me consistently. It never works with the squirrel wrapper though

**Comment by quotidian-ennui** [comment]
> 
> 
> @quotidian-ennui does it work allways if you use "C:\Users<me>\AppData\Local\Fork\app-1.38.0\Fork.exe"? because like that, it works for me consistently. It never works with the squirrel wrapper though

If I spam the `fork` function repeatedly from git-bash, then multiple windows will occur (i.e. fork, up arrow, enter, up arrow, enter etc.) at some point;
This happens with both `/Fork/Fork.exe` and `/Fork/app-1.38.0/Fork.exe`

Generally, I'm not spamming; I imagine that the application hasn't been registered yet (so it's a timing issue), and my spamming means that fork can't find the running instance, so it starts a new one.

**Comment by felixhirt** [comment]
Ah, but this is only when fork is not allready running when you start spamming i presume? In that case it woud make sense i guess (because there is no runnign instance, as you mentioned)

**Comment by felixhirt** [comment]
During testing, i just had the situation where the runnig fork instance switched to the project i told it to open, but a second window opened non the less. Is it possible there is a problem with the IPC response by the running instance to the new one?

**Comment by quotidian-ennui** [comment]
> 
> 
> Ah, but this is only when fork is not allready running when you start spamming i presume? In that case it woud make sense i guess (because there is no runnign instance, as you mentioned)

No, I open fork and "switch to it"; do something;, switch back to bash, `fork` again; (repeat). The first few times, it will show the same instance; but at this point, if I start spamming, then eventually a new fork window appears.  

It smells like a  timing issue because the spamming probably means the IPC can't return quickly enough; so fork just assumes there isn't a running instance it can switch to and creates a new window.
The behaviour goes a bit like this :

```console
xxx@phooey MINGW64 /c/work/dev/amqp (develop)
$ fork
[1] 11044 / No fork open, so new window, switch to it, and do something.

xxx@phooey MINGW64 /c/work/dev/amqp (develop)
$ fork / resuses the existng window, switch to it, and do something.
[2] 11050

xxx@phooey MINGW64 /c/work/dev/amqp (develop)
$ fork / reuses the existing window.
[3] 11056
[2]   Done                    "/c/Users/xxx/AppData/Local/Fork/app-1.38.0/Fork.exe" $(cygpath -m $(pwd))

xx@phooey MINGW64 /c/work/dev/amqp (develop)
$ fork  / now we start spamming...
[4] 11062
[3]   Done                    "/c/Users/xxx/AppData/Local/Fork/app-1.38.0/Fork.exe" $(cygpath -m $(pwd))

xxx@phooey MINGW64 /c/work/dev/amqp (develop)
$ fork
[5] 11068

xxx@phooey MINGW64 /c/work/dev/amqp (develop)
$ fork
[6] 11074

xxx@phooey MINGW64 /c/work/dev/amqp (develop)
$ fork
[7] 11080

xxx@phooey MINGW64 /c/work/dev/amqp (develop)
$ fork
[8] 11086

xxx@phooey MINGW64 /c/work/dev/amqp (develop)
$ fork  / at this point, I have 4 extra windows showing up in the task bar, so I wait a while; and close the other instances...
[9] 11092
[4]   Done                    "/c/Users/xxx/AppData/Local/Fork/app-1.38.0/Fork.exe" $(cygpath -m $(pwd))
[5]   Done                    "/c/Users/xxx/AppData/Local/Fork/app-1.38.0/Fork.exe" $(cygpath -m $(pwd))
[6]   Done                    "/c/Users/xxx/AppData/Local/Fork/app-1.38.0/Fork.exe" $(cygpath -m $(pwd))
[7]   Done                    "/c/Users/xxx/AppData/Local/Fork/app-1.38.0/Fork.exe" $(cygpath -m $(pwd))
[8]   Done                    "/c/Users/xxx/AppData/Local/Fork/app-1.38.0/Fork.exe" $(cygpath -m $(pwd))

xxx@phooey MINGW64 /c/work/dev/amqp (develop)
$ fork / I wait a while, and this will reuse the existing window.
[10] 11098
[9]   Done                    "/c/Users/xxx/AppData/Local/Fork/app-1.38.0/Fork.exe" $(cygpath -m $(pwd))
```

**Comment by felixhirt** [comment]
I had the same impression so far, maybe the timeouts for the ipc are a bit too agressive

**Comment by DanPristupov** [comment]
> I had the same impression so far, maybe the timeouts for the ipc are a bit too agressive

You are right, it's 100ms timeout.

It looks like the only way to solve the problem is to create my own command line runner and put it to some folder which is included in PATH by default. Somewhere like `/usr/local/bin` on unix, but on Windows. What do you think?

**Comment by felixhirt** [comment]
Im not sure i understand. What would you gain by having your own command line runner? Is the problem a side effect of squirrel?

**Comment by DanPristupov** [comment]
@felixhirt yes, Squirrel cuts off the working directory info (https://github.com/ForkIssues/TrackerWin/issues/266#issuecomment-522463736).

Making an own CLI runner is a way to fix this.

**Comment by felixhirt** [comment]
Ah yes thats the other bug, but im not sure these two problems are related? (relative paths to repos / IPC timeouts). Also if you dont use the squirrel wrapper, do you still get the updates etc?

**Comment by DanPristupov** [comment]
No, these are two different problems.

1. Lack of working directory info makes impossible to handle cli command like `fork .` or `fork repo1` (because Fork doesn't know where the 'repo1' repo is located). This will also not allow me to introduce commands like `fork diff` (to open current repo and activate the changes view) or `fork history /path/to/file` (to open current repo and open file history dialog for the given file in repo). 

2. If you run fork with CLI parameters it checks for an other instance with 100ms timeout. However it might happen (?) that the other instance is already busy handling previous request. I can't make the timeout too long, because it will slow down/freeze the startup time.

CLI runner will be used only to handle CLI commands and should not affect Squirrel and the automatic updates.

**Comment by felixhirt** [comment]
That would be a solution for the first problem then, if you plan to add advanced commands like the ones you presented it would be needed yes.
About the IPC: Have you considered (do you allready) check first if a process named "fork.exe" is allready running on the machine? If it does, you can try to communicate with it over ipc, and also be more generous with timeouts etc. because it is much more likely that it will work and you wont have to open a window

**Comment by DanPristupov** [comment]
> About the IPC: Have you considered (do you allready) check first if a process named "fork.exe" is allready running on the machine? If it does, you can try to communicate with it over ipc, and also be more generous with timeouts etc.

This is exactly what I do. I didn't investigate the problem yet, but I suppose that the IPC-server thread can be busy handling previous request if you spam commands to it.

**Comment by michaelakin** [comment]
Here is a powershell function and alias I put in my powershell profile that will load Fork and open the repo in the current folder.

```
Function loadFork { c:\users\user.folder\AppData\Local\Fork\fork.exe $pwd }
Set-Alias -Name fork -Value loadFork
```

**Comment by tdamsma** [comment]
Just a pointer for those wanting to open fork from WSL: path names must be provided in windows format as fork is a windows program. The utility wslpath can be used like so to open fork in the current folder.

```bash
/mnt/c/Users/$WINUSER/AppData/Local/Fork/Fork.exe $(wslpath -w .)

# $WINUSER can be hardcoded or set like so:
export WINUSER=$(pushd /mnt/c > /dev/null && cmd.exe /q /c "echo %USERNAME%" | rev | cut -c 2- | rev ) 
```

In ended up adding this to my shell so I can run `fork .` or `fork some_path` from WSL:

```bash
export WINUSER=$(pushd /mnt/c > /dev/null && cmd.exe /q /c "echo %USERNAME%" | rev | cut -c 2- | rev ) 
alias fork='load_fork() { /mnt/c/Users/$WINUSER/AppData/Local/Fork/Fork.exe $(wslpath -w $@) };load_fork'
```

**Comment by ravindUwU** [comment]
Has the name of the CLI argument been decided by any chance?

VSCode uses `-n/--new-window` and `-r/--reuse-window`. Sublime opens files in the last used window, unless `-n/--new-window` is specified.

I think Fork should use the same terminology to keep things consistent 🤔.
